PWGHF/D2H/Tasks/taskFlowCharmHadrons.cxx

```diff
@@ -335,7 +354,31 @@ struct HfTaskFlowCharmHadrons {
       float scalprodCand = cosNPhi * xQVec + sinNPhi * yQVec;
       float cosDeltaPhi = std::cos(harmonic * (phiCand - evtPl));
 
-      fillThn(massCand, ptCand, cent, cosNPhi, cosDeltaPhi, scalprodCand, outputMl);
+      if constexpr (std::is_same<T1, CandD0Data>::value || std::is_same<T1, CandD0DatawMl>::value) {
+        std::vector<float> outputMlD0 = {-999., -999.};
+        std::vector<float> outputMlD0bar = {-999., -999.};
+
+        if (candidate.isSelD0() >= selectionFlag) {
+          massCand = hfHelper.invMassD0ToPiK(candidate);
+          if constexpr (std::is_same<T1, CandD0DatawMl>::value) {
+            for (unsigned int iclass = 0; iclass < classMl->size(); iclass++)
+              outputMlD0[iclass] = candidate.mlProbD0()[classMl->at(iclass)];
+          }
+          fillThn(massCand, ptCand, cent, cosNPhi, cosDeltaPhi, scalprodCand, outputMlD0);
+        }
+        if (candidate.isSelD0bar() >= selectionFlag) {
+          massCand = hfHelper.invMassD0barToKPi(candidate);
+          if constexpr (std::is_same<T1, CandD0DatawMl>::value) {
+            for (unsigned int iclass = 0; iclass < classMl->size(); iclass++)
+              outputMlD0bar[iclass] = candidate.mlProbD0bar()[classMl->at(iclass)];
+          }
+          fillThn(massCand, ptCand, cent, cosNPhi, cosDeltaPhi, scalprodCand, outputMlD0bar);
+        }
+      } // TODO: whether to put all the mass calculations here
+
+      if constexpr (DecayChannel != DecayChannel::D0ToPiK) {
+        fillThn(massCand, ptCand, cent, cosNPhi, cosDeltaPhi, scalprodCand, outputMl);
+      }
```

Comment:
There exists a case where isSelD0 and isSelD0bar both are 1, called a reflection.
D0 and D0bar have different invMass calculation methods:

-    ` hfHelper.invMassD0ToPiK(candidate)`

-    `hfHelper.invMassD0barToKPi(candidate)`

If both `isSelD0 `and `isSelD0bar` are equal to `1`, then need to be filled twice.
And different mlProb:

-   `candidate.mlProbD0()`

-   `candidate.mlProbD0bar()`

For now, there isn't a need to split them into two processes, just need to be calculated by different `invMass`. Then, I put them after `scalprodCand ` and `cosDeltaPhi`, just for filling.

Reply:
What you write is correct, however it works also if you implement my suggestion: in fact, these "ambiguous" cases will be included in both partitions, and hence they will be filled twice (once in the `D0` case, and once in the `D0bar` case)

The two implementations are equivalent, however it would be good to have the two mass hypotheses handled similarly for all particle species, in order to have it consistent